src/neo/Ledger/Blockchain.cs

```diff
@@ -331,6 +331,7 @@ private VerifyResult OnNewBlock(Block block)
                 if (!block.Hash.Equals(header_index[(int)block.Index]))
                     return VerifyResult.Invalid;
             }
+            block_cache[block.Hash] = block;
```

Comment:
Please use `block_cache.TryAdd()`. It adds the item only if the key doesn't exist.